verifier-py/insights_ansible_playbook_verifier/app.py

```diff
@@ -0,0 +1,238 @@
+        )
+
+    fields: list[str] = playbook.get("vars", {}).get("insights_signature_exclude").split(",")
+    result: dict = copy.deepcopy(playbook)
+
+    for field in fields:
+        elements: list[str] = [string for string in field.split("/") if string != ""]
+        if len(elements) not in (1, 2):
+            raise PreconditionError(
+                f"Variable field '{field}' is too deep or shallow, only one or two levels are allowed."
+            )
+        if elements[0] not in VARIABLE_FIELDS:
+            raise PreconditionError("Variable field '{field}' cannot be excluded.")
+
+        if len(elements) == 1:
+            if elements[0] not in result.keys():
+                raise PreconditionError(
+                    f"Variable field '{field}' is not present in the playbook."
+                )
+            logger.debug(f"Excluding variable field '{field}'.")
+            del result[elements[0]]
+        else:
+            if elements[1] not in result[elements[0]].keys():
+                raise PreconditionError(
+                    f"Variable field '{field}' is not present in the playbook."
+                )
+            logger.debug(f"Excluding variable field '{field}'.")
+            del result[elements[0]][elements[1]]
+
+    return result
+
+
+def create_playbook_digest(playbook: bytes) -> bytes:
+    sha = hashlib.sha256()
+    sha.update(playbook)
+    return sha.digest()
+
+
+def verify_playbook(playbook: dict) -> typing.Tuple[bool, bytes]:
+    """Verify playbook's signature.
+
+    :param playbook: Parsed playbook.
+    :returns: OK?, playbook digest.
+    """
+    logger.info(f"Preparing to verify playbook '{playbook.get('name')}'.")
+
+    b64_signature: str = playbook.get("vars", {}).get("insights_signature", "")
+    if b64_signature == "":
+        raise PreconditionError("The playbook snippet does not contain signature.")
+
+    cleaned_playbook: dict = clean_playbook(playbook)
+    serialized_playbook: str = serialize_playbook(cleaned_playbook)
+    digest: bytes = create_playbook_digest(serialized_playbook.encode("utf-8"))
+    signature: bytes = base64.b64decode(b64_signature)
+
+    gpg_key: bytes = get_gpg_key_content()
+
+    temp_dir = tempfile.TemporaryDirectory(prefix="insights-ansible-playbook-verifier-")
+    digest_file = pathlib.Path(temp_dir.name) / "digest"
+    digest_file.write_bytes(digest)
+    signature_file = pathlib.Path(temp_dir.name) / "signature"
+    signature_file.write_bytes(signature)
+    key_file = pathlib.Path(temp_dir.name) / "key"
+    key_file.write_bytes(gpg_key)
+
+    logger.info(f"Cryptographically verifying playbook '{playbook.get('name')}'.")
+    result: GPGCommandResult = verify_gpg_signed_file(digest_file, signature_file, key_file)
+
+    if not result.ok:
+        logger.error(
+            f"Playbook content failed to match its digest's signature: {serialized_playbook}."
+        )
+
+    temp_dir.cleanup()
+    return result.ok, digest
```

Comment:
returning `(bool, bytes)` feels a bit "go-ish", and even more awkward considering that the function raises exceptions for failures anyway

my suggestion here would be:
- add a new "lightweight" exception similar to `PreconditionError` -- eg `GPGValidationError`
- raise it in case `not result.ok`
- simply return the digest from `verify_playbook()`
- drop the "ok" checks from the callers of `verify_playbook()`

this way, the GPG validation failure will still result in a program failure, and it can be separately caught in callers/tests as needed to check for that specific situation